File: frontend/js/views/expense_reimbursement.js
// Expense reimbursement request generator — itemized expenses + mileage →
// total reimbursement, via /calc/expense-reimbursement. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

const SEED = [
    { desc: 'Hotel', amt: 120 },
    { desc: 'Meals', amt: 50 },
    { desc: 'Parking', amt: 30 },
];

function rowHtml(e) {
    return `
        <div class="mpb-row sdi-row">
            <input type="text" class="er-desc" placeholder="${esc(t('view.er.ph.desc'))}" value="${esc(e.desc || '')}">
            <input type="number" step="0.01" min="0" class="er-amt" placeholder="${esc(t('view.er.ph.amt'))}" value="${e.amt}">
            <button type="button" class="er-del" data-i18n="view.er.remove">Remove</button>
        </div>`;
}

export async function renderExpenseReimbursement(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.er.h1.title">// EXPENSE REIMBURSEMENT</span></h1>
        <p class="muted small" data-i18n="view.er.hint.intro">
            An employee's itemized claim for reimbursement of business expenses plus mileage. It sums the
            itemized expenses, computes the mileage reimbursement (business miles × the mileage rate), and
            totals the two, then assembles the request with an approval line. Drafting aid, not legal/tax
            advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.er.h2.inputs">Request details</h2>
            <form id="er-form" class="inline-form">
                <label><span data-i18n="view.er.label.company">Company</span>
                    <input type="text" name="company_name" value=""></label>
                <label><span data-i18n="view.er.label.employee">Employee</span>
                    <input type="text" name="employee_name" value=""></label>
                <label><span data-i18n="view.er.label.period">Period / purpose</span>
                    <input type="text" name="period" value="June 2026 client trip"></label>
                <label><span data-i18n="view.er.label.miles">Business miles</span>
                    <input type="number" step="0.1" min="0" name="business_miles" value="100"></label>
                <label><span data-i18n="view.er.label.rate">Mileage rate ($/mi)</span>
                    <input type="number" step="0.001" min="0" name="mileage_rate_usd" value="0.67"></label>
                <label><span data-i18n="view.er.label.date">Submitted date</span>
                    <input type="date" name="submitted_date" value="2026-06-30" required></label>
            </form>
            <div class="mpb-head sdi-head">
                <span data-i18n="view.er.col.desc">Expense</span>
                <span data-i18n="view.er.col.amt">Amount ($)</span>
                <span></span>
            </div>
            <div id="er-rows">${SEED.map(rowHtml).join('')}</div>
            <button type="button" id="er-add" class="secondary" data-i18n="view.er.add">+ Add expense</button>
        </div>
        <div id="er-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#er-form');
    const rowsEl = mount.querySelector('#er-rows');

    const generate = async () => {
        const fd = new FormData(form);
        const expenses = [...rowsEl.querySelectorAll('.sdi-row')].map((r) => ({
            description: (r.querySelector('.er-desc').value || '').trim(),
            amount_usd: Number(r.querySelector('.er-amt').value) || 0,
        })).filter((e) => e.description || e.amount_usd);
        const body = {
            company_name: (fd.get('company_name') || '').trim(),
            employee_name: (fd.get('employee_name') || '').trim(),
            period: (fd.get('period') || '').trim(),
            expenses,
            business_miles: Number(fd.get('business_miles')) || 0,
            mileage_rate_usd: Number(fd.get('mileage_rate_usd')) || 0,
            submitted_date: fd.get('submitted_date'),
        };
        try {
            const doc = await api.calcExpenseReimbursement(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.er.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);

    mount.querySelector('#er-add').addEventListener('click', () => {
        rowsEl.insertAdjacentHTML('beforeend', rowHtml({ desc: '', amt: '' }));
        applyUiI18n(rowsEl.lastElementChild);
        generate();
    });
    rowsEl.addEventListener('click', (e) => {
        if (e.target.classList.contains('er-del')) {
            e.target.closest('.sdi-row').remove();
            generate();
        }
    });
    form.addEventListener('input', () => { live(); });
    rowsEl.addEventListener('input', () => { live(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#er-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.er.card.total">Total reimbursement</div>
                    <div class="value">${money(doc.total_reimbursement_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.er.card.expenses">Expenses</div>
                    <div class="value">${money(doc.expenses_total_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.er.card.mileage">Mileage</div>
                    <div class="value">${money(doc.mileage_reimbursement_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="er-copy" type="button" data-i18n="view.er.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="er-download" type="button" data-i18n="view.er.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#er-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.er.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.er.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#er-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'expense-reimbursement.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/expense_reimbursement.rs
//! Expense reimbursement request — an employee's itemized claim for
//! reimbursement of business expenses plus mileage. It sums the itemized
//! expenses, computes the mileage reimbursement (business miles × the mileage
//! rate), and totals the two, then assembles the request with an approval line.
//! Drafting aid, not legal/tax advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct ExpenseItem {
    pub description: String,
    pub amount_usd: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ReimbursementInput {
    pub company_name: String,
    pub employee_name: String,
    /// Period or purpose the expenses cover (free text).
    pub period: String,
    #[serde(default)]
    pub expenses: Vec<ExpenseItem>,
    #[serde(default)]
    pub business_miles: f64,
    /// Reimbursement rate per mile (e.g. the IRS standard mileage rate).
    #[serde(default)]
    pub mileage_rate_usd: f64,
    pub submitted_date: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ReimbursementRequest {
    pub title: String,
    pub expenses_total_usd: f64,
    pub mileage_reimbursement_usd: f64,
    pub total_reimbursement_usd: f64,
    pub item_count: usize,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &ReimbursementInput) -> ReimbursementRequest {
    let expenses_total = cents(i.expenses.iter().map(|e| e.amount_usd).sum());
    let mileage_reimbursement = cents(i.business_miles * i.mileage_rate_usd);
    let total = cents(expenses_total + mileage_reimbursement);

    let items_body = if i.expenses.is_empty() {
        "No itemized expenses submitted.".to_string()
    } else {
        let mut lines: Vec<String> = i
            .expenses
            .iter()
            .map(|e| format!("  • {} — {}", e.description, money(e.amount_usd)))
            .collect();
        lines.push(format!("Itemized subtotal: {}", money(expenses_total)));
        lines.join("\n")
    };

    let mileage_body = if i.business_miles > 0.0 {
        format!(
            "{:.1} business miles at {} per mile = {}.",
            i.business_miles,
            money(i.mileage_rate_usd),
            money(mileage_reimbursement)
        )
    } else {
        "No mileage claimed.".to_string()
    };

    let clauses = vec![
        DocClause {
            heading: "Header".into(),
            body: format!(
                "Employee: {}\nCompany: {}\nPeriod / purpose: {}\nSubmitted: {}",
                i.employee_name, i.company_name, i.period, i.submitted_date
            ),
        },
        DocClause { heading: "1. Itemized Expenses".into(), body: items_body },
        DocClause { heading: "2. Mileage".into(), body: mileage_body },
        DocClause {
            heading: "3. Total Reimbursement".into(),
            body: format!(
                "Total reimbursement requested: {} (expenses {} + mileage {}).",
                money(total),
                money(expenses_total),
                money(mileage_reimbursement)
            ),
        },
        DocClause {
            heading: "4. Certification".into(),
            body: "I certify that the expenses above were incurred for legitimate business purposes and that receipts are attached where required by company policy.".into(),
        },
        DocClause {
            heading: "Approval".into(),
            body: format!(
                "Employee: ____________________  Date: {}\n{}\n\nApproved by: ____________________  Date: __________",
                i.submitted_date, i.employee_name
            ),
        },
    ];

    ReimbursementRequest {
        title: "Expense Reimbursement Request".into(),
        expenses_total_usd: expenses_total,
        mileage_reimbursement_usd: mileage_reimbursement,
        total_reimbursement_usd: total,
        item_count: i.expenses.len(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn item(d: &str, a: f64) -> ExpenseItem {
        ExpenseItem { description: d.into(), amount_usd: a }
    }

    fn base() -> ReimbursementInput {
        ReimbursementInput {
            company_name: "Acme Inc".into(),
            employee_name: "Dana Road".into(),
            period: "June 2026 client trip".into(),
            expenses: vec![item("Hotel", 120.0), item("Meals", 50.0), item("Parking", 30.0)],
            business_miles: 100.0,
            mileage_rate_usd: 0.67,
            submitted_date: "2026-06-30".into(),
        }
    }

    #[test]
    fn totals() {
        let d = generate(&base());
        assert!(close(d.expenses_total_usd, 200.0));
        assert!(close(d.mileage_reimbursement_usd, 67.0));
        assert!(close(d.total_reimbursement_usd, 267.0));
        assert_eq!(d.item_count, 3);
    }

    #[test]
    fn itemized_lines_and_subtotal() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading.contains("Itemized")).unwrap();
        assert!(c.body.contains("Hotel — $120.00"));
        assert!(c.body.contains("Itemized subtotal: $200.00"));
    }

    #[test]
    fn mileage_clause() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading == "2. Mileage").unwrap();
        assert!(c.body.contains("100.0 business miles at $0.67 per mile = $67.00"));
    }

    #[test]
    fn no_mileage_states_none() {
        let d = generate(&ReimbursementInput { business_miles: 0.0, ..base() });
        assert!(close(d.mileage_reimbursement_usd, 0.0));
        assert!(close(d.total_reimbursement_usd, 200.0));
        let c = d.clauses.iter().find(|c| c.heading == "2. Mileage").unwrap();
        assert!(c.body.contains("No mileage"));
    }

    #[test]
    fn no_expenses_states_none() {
        let d = generate(&ReimbursementInput { expenses: vec![], ..base() });
        assert!(close(d.expenses_total_usd, 0.0));
        let c = d.clauses.iter().find(|c| c.heading.contains("Itemized")).unwrap();
        assert!(c.body.contains("No itemized expenses"));
    }

    #[test]
    fn total_clause_breaks_down() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading.contains("Total Reimbursement")).unwrap();
        assert!(c.body.contains("$267.00"));
        assert!(c.body.contains("expenses $200.00 + mileage $67.00"));
    }
}
